C7/Map/MapView.cs

```diff
@@ -34,17 +36,96 @@ public partial class MapView : Node2D {
 
-		public override void _Draw() {
-			GD.Print("draw...");
-			game.animTracker.update();
-			foreach ((string id, AnimationTracker.ActiveAnimation anim) in game.animTracker.activeAnims) {
-				GD.Print($"{id}: {anim.ToString()}");
+		private Dictionary<MapUnit, UnitSprite> unitSprites = new Dictionary<MapUnit, UnitSprite>();
+		private CursorSprite cursor;
+
+		private UnitSprite spriteFor(MapUnit unit) {
+			UnitSprite sprite = unitSprites.GetValueOrDefault(unit, null);
+			if (sprite is null) {
+				sprite = new UnitSprite(game.civ3AnimData);
+				unitSprites.Add(unit, sprite);
+				AddChild(sprite);
 			}
-			base._Draw();
+			return sprite;
+		}
+
+		private Vector2 getSpriteLocalPosition(Tile tile, MapUnit.Appearance appearance) {
+			Vector2 position = tilemap.MapToLocal(stackedCoords(tile));
+			Vector2 offset = tileSize * new Vector2(appearance.offsetX, appearance.offsetY) / 2;
+			return position + offset;
 		}
 
-		public override void _Process(double delta) {
-			base._Process(delta);
+		private void animateUnit(Tile tile, MapUnit unit) {
+			// TODO: simplify AnimationManager and drawing animations it is unnecessarily complex
+			// - also investigate if the custom offset tracking and SetFrame can be replaced by
+			//   engine functionality
+			MapUnit.Appearance appearance = game.animTracker.getUnitAppearance(unit);
+			string name = AnimationManager.AnimationKey(unit.unitType, appearance.action, appearance.direction);
+			C7Animation animation = game.civ3AnimData.forUnit(unit.unitType, appearance.action);
+			animation.loadSpriteAnimation();
+			UnitSprite sprite = spriteFor(unit);
+			int frame = sprite.GetNextFrameByProgress(name, appearance.progress);
+			float yOffset = sprite.FrameSize(name).Y / 4f; // TODO: verify actual value
+			Vector2 position = getSpriteLocalPosition(tile, appearance);
+			sprite.Position = position - new Vector2(0, yOffset);
+			Color civColor = new Color(unit.owner.color);
+			sprite.SetColor(civColor);
+			sprite.SetAnimation(name);
+			sprite.SetFrame(frame);
+			sprite.Show();
+
+			if (unit == game.CurrentlySelectedUnit) {
+				cursor.Position = position;
+				cursor.Show();
+			}
+		}
+
+		private MapUnit selectUnitToDisplay(List<MapUnit> units) {
+			if (units.Count == 0) {
+				return MapUnit.NONE;
+			}
+			MapUnit bestDefender = units[0], selected = null, interesting = null;
+			MapUnit currentlySelected = game.CurrentlySelectedUnit;
+			foreach (MapUnit unit in units) {
+				if (unit == currentlySelected) {
+					selected = unit;
+				}
+				if (unit.HasPriorityAsDefender(bestDefender, currentlySelected)) {
+					bestDefender = unit;
+				}
+				if (game.animTracker.getUnitAppearance(unit).DeservesPlayerAttention()) {
+					interesting = unit;
+				}
+			}
+			// Prefer showing the selected unit, secondly show one doing a relevant animation, otherwise show the top defender
+			return selected is not null ? selected : (interesting is not null ? interesting : bestDefender);
```

Comment:
you can condense this with the null coalescing operator:
`return selected ?? interesting ?? bestDefender`